jep/401/README.adoc

```diff
@@ -0,0 +1,246 @@
+
+=== Jenkins core
+
+* Let’s consider the following definition of the `Header` on: `core/src/main/java/jenkins/views/Header.java`
+
+```
+package jenkins.views;
+
+import hudson.ExtensionPoint;
+
+public abstract class Header implements ExtensionPoint {
+
+    /**
+     * Check if the header is enabled. By default it is if installed,
+     * but the logic is deferred in the plugins.
+     * @return
+     */
+    public abstract boolean isHeaderEnabled();
+
+}
+```
+
+* Let’s consider the following implementation of the Jenkins header on: `core/src/main/java/jenkins/views/JenkinsHeader.java`
+
+```
+package jenkins.views;
+
+import hudson.Extension;
+
+@Extension(ordinal = Integer.MIN_VALUE)
+public class JenkinsHeader extends Header {
+
+    @Override
+    public boolean isHeaderEnabled() {
+        return true;
+    }
+}
+```
+
+* As mentioned before, the `Functions` method `header()` will retrieve the available headers via `Extension` lookup that are enabled and for those obtained it will provide the one with max priority (ordinal value)
+
+```
+[...]
+@Restricted(NoExternalUse.class)
+@CheckForNull
+public static Header header() {
+    List<Header> headers = ExtensionList.lookup(Header.class).stream()
+                .filter(header -> header.isHeaderEnabled())
+                .collect(Collectors.toList());
+    if (headers.size() > 0) {
+        if (headers.size() > 1) {
+            LOGGER.warning("More than one configured header. This should not happen. Serving the Jenkins default header and please review");
+        } else {
+            return headers.get(0);
+        }
+    }
+    return new JenkinsHeader();
+}
+```
+
+* Once we launch Jenkins with the proposed changes on the core, we will obtain the expected/current header working without any issue
+
+=== Custom UI plugin
+
+* Create a new plugin following the usual procedure
+* Provide an implementation of the custom Header (i.e: `custom-ui-plugin/src/main/java/foo/bar/CustomHeader.java`)
+
+```
+[...]
+@Extension(optional = true, ordinal = Integer.MAX_VALUE)
+public class CustomHeader extends Header {
+
+    @Override
+    public boolean isHeaderEnabled() {
+        // Disable/enable the header based on an ENV var
+        boolean isDisabled = "true".equalsIgnoreCase(System.getenv("CUSTOM_HEADER_DISABLE"));
+        return !isDisabled;
+    }
+}
+```
+
+* Use an `ExtensionFilter` to avoid other `Header` to be available (i.e.: `custom-ui-plugin/src/main/java/foo/bar/CustomHeaderExtensionFilter.java`)
+
+```
+[...]
+@Extension
+public class CustomHeaderExtensionFilter extends ExtensionFilter {
+
+    @Override
+    public <T> boolean allows(Class<T> type, ExtensionComponent<T> component) {
+        boolean isCustom = component.getInstance().getClass().equals(CustomHeader.class);
+        boolean isForbiddenHeader = type.equals(Header.class) && !isCustom;
+        return !isForbiddenHeader;
+    }
+}
+```
```

Comment:
ditto: dangerous to share what we plan for the header

Reply:
Same than previous. I will remove this.

Reply:
Removed in https://github.com/imonteroperez/jep/pull/1/commits/b240f8f85965918e0fd775ecfad4fe2121fe555b